cli/dcoscli/tail/main.py

```diff
@@ -0,0 +1,369 @@
+        unreachable, we stop trying to read from it.
+    :rtype: (str, [MesosFile])
+    """
+
+    reachable_files = list(mesos_files)
+
+    # TODO switch to map
+    for job, mesos_file in _stream(fn, mesos_files):
+        try:
+            lines = job.result()
+        except DCOSException as e:
+            # The read function might throw an exception if read.json
+            # is unavailable, or if the file doesn't exist in the
+            # sandbox.  In any case, we silently remove the file and
+            # continue.
+            logger.warning("Error reading file: {}".format(e))
+
+            reachable_files.remove(mesos_file)
+            continue
+
+        curr_header = _output(curr_header,
+                              len(reachable_files) > 1,
+                              str(mesos_file),
+                              lines)
+
+    return curr_header, reachable_files
+
+
+def _output(curr_header, output_header, header, lines):
+    """Prints a sequence of lines.  If `header` is different than
+    `curr_header`, first print the header.
+
+    :param curr_header: most recently printed header
+    :type curr_header: str
+    :param output_header: whether or not to output the header
+    :type output_header: bool
+    :param header: header for `lines`
+    :type header: str
+    :param lines: lines to print
+    :type lines: [str]
+    :returns: `header`
+    :rtype: str
+    """
+
+    if lines:
+        if output_header and header != curr_header:
+            emitter.publish('===> {} <==='.format(header))
+        for line in lines:
+            emitter.publish(line)
+    return header
+
+
+def _stream(fn, objs):
+    """Apply `fn` to `objs` in parallel, yielding the (Future, obj) for
+    each as it completes.
+
+    :param fn: function
+    :type fn: function
+    :param objs: objs
+    :type objs: objs
+    :returns: iterator over (Future, typeof(obj))
+    :rtype: iterator over (Future, typeof(obj))
+
+    """
+
+    with concurrent.futures.ThreadPoolExecutor(20) as pool:
+        jobs = {pool.submit(fn, obj): obj for obj in objs}
+        for job in concurrent.futures.as_completed(jobs):
+            yield job, jobs[job]
+
+
+# A liberal estimate of a line size.  Used to estimate how much data
+# we need to fetch from a file when we want to read N lines.
+LINE_SIZE = 200
+
+
+def _read_last_lines(num_lines, mesos_file):
+    """Returns the last `num_lines` of a file, or less if the file is
+    smaller.  Seeks to EOF.
+
+    :param num_lines: number of lines to read
+    :type num_lines: int
+    :param mesos_file: file to read
+    :type mesos_file: MesosFile
+    :returns: lines read
+    :rtype: [str]
+    """
+
+    file_size = mesos_file.size()
+
+    # estimate how much data we need to fetch to read `num_lines`.
+    fetch_size = LINE_SIZE * num_lines
+
+    end = file_size
+    start = max(file_size - fetch_size, 0)
+    data = ''
+    while True:
```

Comment:
can we do `while (len(lines) <= num_lines or start == 0)`... I don't like `break` statements